### DISCUSSION TITLE
Custom training Entity Linker component

### DISCUSSION THREAD

**Discussion by shrinidhin** [explosion/spaCy#10713] [Help: Coding & Implementations] [answered]
Hello. So I am working on building an Entity Linker component, to use with my custom trained NER. I have gone through the documentation, the video tutorial along with the ner-emersons example project. I had a few doubts about the same: 

1. Is it required to add all possible aliases of the entity to the knowledge base?I have a few aliases for each entity which I can add, However Shouldn't this be trainable?I mean since I am already tagging the KB_IDs in the training data for different ways the entity is referenced, shouldn't the entity linker learn these names on training?If not, then how is this trainable component different from a simple python script which matches aliases from a locally saved files?
2. Are aliases added to the Knowledge base case sensitive?For eg: Will `Nifty 50` and `NIFTY 50` be recognized as two different aliases for the same entity?

Any inputs will be helpful. Thank you!

**Comment by polm** [comment, marked as answer]
> Is it required to add all possible aliases of the entity to the knowledge base?I have a few aliases for each entity which I can add, However Shouldn't this be trainable?

The current implementation of the Entity Linker requires this. However, the candidate generation is controlled by a function, so you can customize that function if you want. For example, rather than requiring exact matches, you could use edit distance or something.

> since I am already tagging the KB_IDs in the training data for different ways the entity is referenced, shouldn't the entity linker learn these names on training?

If you don't already have a list of aliases, it can be useful to create one by extracting values from data you've manually annotated.

> If not, then how is this trainable component different from a simple python script which matches aliases from a locally saved files?

If most of your aliases are unambiguous - they refer to only one entity - then there won't be a meaningful difference. The NEL component is useful in situations like the Emerson examples in the demo project. The component learns to distinguish between different things with the same name. Learning different names for the same thing is a related but distinct problem.

> Are aliases added to the Knowledge base case sensitive?For eg: Will Nifty 50 and NIFTY 50 be recognized as two different aliases for the same entity?

As stated [in the documentation](https://spacy.io/api/architectures/#CandidateGenerator) (and I think you found out by experimentation), the default candidate generator is case-sensitive. You can define a custom case generator that's case insensitive by copying the code for the default one and modifying it slightly.

**Reply by shrinidhin** [reply]
@polm Thank you!
